File: src/commonTest/kotlin/com/scoreo/application/GetHeadToHeadUseCaseTest.kt
package com.scoreo.application

import com.scoreo.infrastructure.InMemoryGameTypeRepository
import com.scoreo.infrastructure.InMemoryMatchRepository
import com.scoreo.infrastructure.InMemoryPlayerRepository
import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.Match
import com.scoreo.domain.model.Player
import com.scoreo.domain.model.PlayerScore
import com.scoreo.domain.model.WinCondition
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

private const val ELO_INITIAL = 1200
private const val ELO_K = 32
private val ELO_AFTER_WIN = ELO_INITIAL + (ELO_K * 0.5).toInt()   // 1216
private val ELO_AFTER_LOSS = ELO_INITIAL - (ELO_K * 0.5).toInt()  // 1184

class GetHeadToHeadUseCaseTest {

    private fun buildUseCase(
        matchRepo: InMemoryMatchRepository = InMemoryMatchRepository(),
        gameTypeRepo: InMemoryGameTypeRepository = InMemoryGameTypeRepository(),
        playerRepo: InMemoryPlayerRepository = InMemoryPlayerRepository(),
    ) = GetHeadToHeadUseCase(matchRepo, gameTypeRepo, playerRepo)

    @Test
    fun `empty when no matches`() {
        val useCase = buildUseCase()
        assertTrue(useCase().isEmpty())
    }

    @Test
    fun `returns player detail with total wins and losses`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice"))
            it.save(Player("p2", "Bob"))
        }
        val gameTypeRepo = InMemoryGameTypeRepository().also {
            it.save(GameType("gt1", "Test", WinCondition.HIGHEST_SCORE))
        }
        val matchRepo = InMemoryMatchRepository().also {
            it.save(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5))))
        }
        val useCase = buildUseCase(matchRepo, gameTypeRepo, playerRepo)

        val result = useCase()

        assertEquals(2, result.size)
        val alice = result.find { it.playerId == "p1" }
        assertEquals(1, alice?.wins)
        assertEquals(0, alice?.losses)
    }

    @Test
    fun `leaderboard sorted by elo descending`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice"))
            it.save(Player("p2", "Bob"))
            it.save(Player("p3", "Charlie"))
        }
        val gameTypeRepo = InMemoryGameTypeRepository().also {
            it.save(GameType("gt1", "Test", WinCondition.HIGHEST_SCORE))
        }
        val matchRepo = InMemoryMatchRepository().also {
            it.save(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5))))
            it.save(Match("m2", 2000L, "gt1", listOf(PlayerScore("p1", 8), PlayerScore("p2", 12))))
            it.save(Match("m3", 3000L, "gt1", listOf(PlayerScore("p3", 10), PlayerScore("p1", 3))))
        }
        val useCase = buildUseCase(matchRepo, gameTypeRepo, playerRepo)

        val result = useCase()

        assertEquals(listOf("p3", "p2", "p1"), result.map { it.playerId })
    }

    @Test
    fun `excludes players with zero matches`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice"))
            it.save(Player("p2", "Bob"))
        }
        val useCase = buildUseCase(playerRepo = playerRepo)

        assertTrue(useCase().isEmpty())
    }

    @Test
    fun `head to head computed correctly`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice"))
            it.save(Player("p2", "Bob"))
        }
        val gameTypeRepo = InMemoryGameTypeRepository().also {
            it.save(GameType("gt1", "Test", WinCondition.HIGHEST_SCORE))
        }
        val matchRepo = InMemoryMatchRepository().also {
            it.save(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5))))
            it.save(Match("m2", 2000L, "gt1", listOf(PlayerScore("p1", 8), PlayerScore("p2", 12))))
            it.save(Match("m3", 3000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 6))))
        }
        val useCase = buildUseCase(matchRepo, gameTypeRepo, playerRepo)

        val result = useCase()
        val alice = result.find { it.playerId == "p1" }!!
        val bob = result.find { it.playerId == "p2" }!!

        assertEquals(1, alice.headToHead.size)
        assertEquals(2, alice.headToHead.first().wins)
        assertEquals(1, alice.headToHead.first().losses)
        assertEquals(1, bob.headToHead.size)
        assertEquals(1, bob.headToHead.first().wins)
        assertEquals(2, bob.headToHead.first().losses)
    }

    @Test
    fun `head to head in multi player match`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice"))
            it.save(Player("p2", "Bob"))
            it.save(Player("p3", "Charlie"))
        }
        val gameTypeRepo = InMemoryGameTypeRepository().also {
            it.save(GameType("gt1", "Test", WinCondition.HIGHEST_SCORE))
        }
        val matchRepo = InMemoryMatchRepository().also {
            it.save(Match("m1", 1000L, "gt1", listOf(
                PlayerScore("p1", 10), PlayerScore("p2", 5), PlayerScore("p3", 3)
            )))
        }
        val useCase = buildUseCase(matchRepo, gameTypeRepo, playerRepo)

        val result = useCase()
        val alice = result.find { it.playerId == "p1" }!!
        val bob = result.find { it.playerId == "p2" }!!
        val charlie = result.find { it.playerId == "p3" }!!

        assertEquals(2, alice.headToHead.size)
        assertEquals(1, alice.headToHead.find { it.opponentId == "p2" }?.wins)
        assertEquals(1, alice.headToHead.find { it.opponentId == "p3" }?.wins)
        assertEquals(1, bob.headToHead.find { it.opponentId == "p1" }?.losses)
        assertEquals(1, charlie.headToHead.find { it.opponentId == "p1" }?.losses)
    }

    @Test
    fun `includes inactive players in results`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice", active = false))
            it.save(Player("p2", "Bob"))
        }
        val gameTypeRepo = InMemoryGameTypeRepository().also {
            it.save(GameType("gt1", "Test", WinCondition.HIGHEST_SCORE))
        }
        val matchRepo = InMemoryMatchRepository().also {
            it.save(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5))))
        }
        val useCase = buildUseCase(matchRepo, gameTypeRepo, playerRepo)

        val result = useCase()

        assertEquals(2, result.size)
        assertTrue(result.any { it.playerId == "p1" })
    }

    @Test
    fun `elo is 1200 for players with no matches`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice"))
        }
        val useCase = buildUseCase(playerRepo = playerRepo)

        assertTrue(useCase().isEmpty())
    }

    @Test
    fun `elo changes after a match`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice"))
            it.save(Player("p2", "Bob"))
        }
        val gameTypeRepo = InMemoryGameTypeRepository().also {
            it.save(GameType("gt1", "Test", WinCondition.HIGHEST_SCORE))
        }
        val matchRepo = InMemoryMatchRepository().also {
            it.save(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5))))
        }
        val useCase = buildUseCase(matchRepo, gameTypeRepo, playerRepo)

        val result = useCase()
        val alice = result.find { it.playerId == "p1" }!!
        val bob = result.find { it.playerId == "p2" }!!

        assertEquals(ELO_AFTER_WIN, alice.elo)
        assertEquals(ELO_AFTER_LOSS, bob.elo)
    }

    @Test
    fun `filters by gameTypeId`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice"))
            it.save(Player("p2", "Bob"))
        }
        val gameTypeRepo = InMemoryGameTypeRepository().also {
            it.save(GameType("gt1", "Type A", WinCondition.HIGHEST_SCORE))
            it.save(GameType("gt2", "Type B", WinCondition.HIGHEST_SCORE))
        }
        val matchRepo = InMemoryMatchRepository().also {
            it.save(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5))))
            it.save(Match("m2", 2000L, "gt2", listOf(PlayerScore("p1", 3), PlayerScore("p2", 10))))
        }
        val useCase = buildUseCase(matchRepo, gameTypeRepo, playerRepo)

        val gt1 = useCase(gameTypeId = "gt1")
        assertEquals(ELO_AFTER_WIN, gt1.find { it.playerId == "p1" }?.elo)
        assertEquals(ELO_AFTER_LOSS, gt1.find { it.playerId == "p2" }?.elo)

        val gt2 = useCase(gameTypeId = "gt2")
        assertEquals(ELO_AFTER_LOSS, gt2.find { it.playerId == "p1" }?.elo)
        assertEquals(ELO_AFTER_WIN, gt2.find { it.playerId == "p2" }?.elo)

        val empty = useCase(gameTypeId = "nonexistent")
        assertTrue(empty.isEmpty())
    }

    @Test
    fun `elo with multiple matches`() {
        val playerRepo = InMemoryPlayerRepository().also {
            it.save(Player("p1", "Alice"))
            it.save(Player("p2", "Bob"))
            it.save(Player("p3", "Charlie"))
        }
        val gameTypeRepo = InMemoryGameTypeRepository().also {
            it.save(GameType("gt1", "Test", WinCondition.HIGHEST_SCORE))
        }
        val matchRepo = InMemoryMatchRepository().also {
            it.save(Match("m1", 1000L, "gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5))))
            it.save(Match("m2", 2000L, "gt1", listOf(PlayerScore("p1", 8), PlayerScore("p2", 12))))
            it.save(Match("m3", 3000L, "gt1", listOf(PlayerScore("p3", 10), PlayerScore("p1", 3))))
        }
        val useCase = buildUseCase(matchRepo, gameTypeRepo, playerRepo)

        val result = useCase()
        val alice = result.find { it.playerId == "p1" }!!
        val bob = result.find { it.playerId == "p2" }!!
        val charlie = result.find { it.playerId == "p3" }!!

        assertTrue(alice.elo < bob.elo)
        assertTrue(bob.elo < charlie.elo)
        assertEquals(listOf("p3", "p2", "p1"), result.map { it.playerId })
    }
}

File: src/application/getHeadToHeadUseCase.test.ts
import { describe, expect, it } from 'vitest'
import type { GameType } from '../domain/model/gameType'
import type { Match } from '../domain/model/match'
import type { Player } from '../domain/model/player'
import { InMemoryGameTypeRepository } from '../infrastructure/testing/inMemoryGameTypeRepository'
import { InMemoryMatchRepository } from '../infrastructure/testing/inMemoryMatchRepository'
import { InMemoryPlayerRepository } from '../infrastructure/testing/inMemoryPlayerRepository'
import { GetHeadToHeadUseCase } from './getHeadToHeadUseCase'

const ELO_INITIAL = 1200
const ELO_K = 32
const ELO_AFTER_WIN = ELO_INITIAL + Math.round(ELO_K * 0.5) // 1216
const ELO_AFTER_LOSS = ELO_INITIAL - Math.round(ELO_K * 0.5) // 1184

function player(id: string, name: string, active = true): Player {
  return { id, name, active }
}

function gameType(id: string, name: string, winCondition: GameType['winCondition']): GameType {
  return {
    id,
    name,
    winCondition,
    tieBreakRule: 'NONE',
    tieBreakCondition: 'HIGHEST_SCORE',
    tieBreakLabel: null,
    active: true,
  }
}

function match(id: string, date: number, gameTypeId: string, playerScores: Match['playerScores'], overrides: Partial<Match> = {}): Match {
  return { id, date, gameTypeId, playerScores, manualWinners: [], secondaryPlayerScores: [], ...overrides }
}

function buildUseCase(
  matchRepo = new InMemoryMatchRepository(),
  gameTypeRepo = new InMemoryGameTypeRepository(),
  playerRepo = new InMemoryPlayerRepository(),
) {
  return new GetHeadToHeadUseCase(matchRepo, gameTypeRepo, playerRepo)
}

describe('GetHeadToHeadUseCase', () => {
  it('empty when no matches', () => {
    expect(buildUseCase().invoke()).toEqual([])
  })

  it('returns player detail with total wins and losses', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice'))
    playerRepo.save(player('p2', 'Bob'))
    const gameTypeRepo = new InMemoryGameTypeRepository()
    gameTypeRepo.save(gameType('gt1', 'Test', 'HIGHEST_SCORE'))
    const matchRepo = new InMemoryMatchRepository()
    matchRepo.save(
      match('m1', 1000, 'gt1', [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 5 },
      ]),
    )
    const result = buildUseCase(matchRepo, gameTypeRepo, playerRepo).invoke()

    expect(result).toHaveLength(2)
    const alice = result.find((r) => r.playerId === 'p1')
    expect(alice?.wins).toBe(1)
    expect(alice?.losses).toBe(0)
  })

  it('leaderboard sorted by elo descending', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice'))
    playerRepo.save(player('p2', 'Bob'))
    playerRepo.save(player('p3', 'Charlie'))
    const gameTypeRepo = new InMemoryGameTypeRepository()
    gameTypeRepo.save(gameType('gt1', 'Test', 'HIGHEST_SCORE'))
    const matchRepo = new InMemoryMatchRepository()
    matchRepo.save(match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }]))
    matchRepo.save(match('m2', 2000, 'gt1', [{ playerId: 'p1', score: 8 }, { playerId: 'p2', score: 12 }]))
    matchRepo.save(match('m3', 3000, 'gt1', [{ playerId: 'p3', score: 10 }, { playerId: 'p1', score: 3 }]))

    const result = buildUseCase(matchRepo, gameTypeRepo, playerRepo).invoke()

    expect(result.map((r) => r.playerId)).toEqual(['p3', 'p2', 'p1'])
  })

  it('excludes players with zero matches', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice'))
    playerRepo.save(player('p2', 'Bob'))
    expect(buildUseCase(undefined, undefined, playerRepo).invoke()).toEqual([])
  })

  it('head to head computed correctly', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice'))
    playerRepo.save(player('p2', 'Bob'))
    const gameTypeRepo = new InMemoryGameTypeRepository()
    gameTypeRepo.save(gameType('gt1', 'Test', 'HIGHEST_SCORE'))
    const matchRepo = new InMemoryMatchRepository()
    matchRepo.save(match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }]))
    matchRepo.save(match('m2', 2000, 'gt1', [{ playerId: 'p1', score: 8 }, { playerId: 'p2', score: 12 }]))
    matchRepo.save(match('m3', 3000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 6 }]))

    const result = buildUseCase(matchRepo, gameTypeRepo, playerRepo).invoke()
    const alice = result.find((r) => r.playerId === 'p1')!
    const bob = result.find((r) => r.playerId === 'p2')!

    expect(alice.headToHead).toHaveLength(1)
    expect(alice.headToHead[0].wins).toBe(2)
    expect(alice.headToHead[0].losses).toBe(1)
    expect(bob.headToHead).toHaveLength(1)
    expect(bob.headToHead[0].wins).toBe(1)
    expect(bob.headToHead[0].losses).toBe(2)
  })

  it('head to head in multi player match', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice'))
    playerRepo.save(player('p2', 'Bob'))
    playerRepo.save(player('p3', 'Charlie'))
    const gameTypeRepo = new InMemoryGameTypeRepository()
    gameTypeRepo.save(gameType('gt1', 'Test', 'HIGHEST_SCORE'))
    const matchRepo = new InMemoryMatchRepository()
    matchRepo.save(
      match('m1', 1000, 'gt1', [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 5 },
        { playerId: 'p3', score: 3 },
      ]),
    )

    const result = buildUseCase(matchRepo, gameTypeRepo, playerRepo).invoke()
    const alice = result.find((r) => r.playerId === 'p1')!
    const bob = result.find((r) => r.playerId === 'p2')!
    const charlie = result.find((r) => r.playerId === 'p3')!

    expect(alice.headToHead).toHaveLength(2)
    expect(alice.headToHead.find((h) => h.opponentId === 'p2')?.wins).toBe(1)
    expect(alice.headToHead.find((h) => h.opponentId === 'p3')?.wins).toBe(1)
    expect(bob.headToHead.find((h) => h.opponentId === 'p1')?.losses).toBe(1)
    expect(charlie.headToHead.find((h) => h.opponentId === 'p1')?.losses).toBe(1)
  })

  it('includes inactive players in results', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice', false))
    playerRepo.save(player('p2', 'Bob'))
    const gameTypeRepo = new InMemoryGameTypeRepository()
    gameTypeRepo.save(gameType('gt1', 'Test', 'HIGHEST_SCORE'))
    const matchRepo = new InMemoryMatchRepository()
    matchRepo.save(match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }]))

    const result = buildUseCase(matchRepo, gameTypeRepo, playerRepo).invoke()

    expect(result).toHaveLength(2)
    expect(result.some((r) => r.playerId === 'p1')).toBe(true)
  })

  it('elo is 1200 for players with no matches', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice'))
    expect(buildUseCase(undefined, undefined, playerRepo).invoke()).toEqual([])
  })

  it('elo changes after a match', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice'))
    playerRepo.save(player('p2', 'Bob'))
    const gameTypeRepo = new InMemoryGameTypeRepository()
    gameTypeRepo.save(gameType('gt1', 'Test', 'HIGHEST_SCORE'))
    const matchRepo = new InMemoryMatchRepository()
    matchRepo.save(match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }]))

    const result = buildUseCase(matchRepo, gameTypeRepo, playerRepo).invoke()
    const alice = result.find((r) => r.playerId === 'p1')!
    const bob = result.find((r) => r.playerId === 'p2')!

    expect(alice.elo).toBe(ELO_AFTER_WIN)
    expect(bob.elo).toBe(ELO_AFTER_LOSS)
  })

  it('filters by gameTypeId', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice'))
    playerRepo.save(player('p2', 'Bob'))
    const gameTypeRepo = new InMemoryGameTypeRepository()
    gameTypeRepo.save(gameType('gt1', 'Type A', 'HIGHEST_SCORE'))
    gameTypeRepo.save(gameType('gt2', 'Type B', 'HIGHEST_SCORE'))
    const matchRepo = new InMemoryMatchRepository()
    matchRepo.save(match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }]))
    matchRepo.save(match('m2', 2000, 'gt2', [{ playerId: 'p1', score: 3 }, { playerId: 'p2', score: 10 }]))
    const useCase = buildUseCase(matchRepo, gameTypeRepo, playerRepo)

    const gt1 = useCase.invoke('gt1')
    expect(gt1.find((r) => r.playerId === 'p1')?.elo).toBe(ELO_AFTER_WIN)
    expect(gt1.find((r) => r.playerId === 'p2')?.elo).toBe(ELO_AFTER_LOSS)

    const gt2 = useCase.invoke('gt2')
    expect(gt2.find((r) => r.playerId === 'p1')?.elo).toBe(ELO_AFTER_LOSS)
    expect(gt2.find((r) => r.playerId === 'p2')?.elo).toBe(ELO_AFTER_WIN)

    expect(useCase.invoke('nonexistent')).toEqual([])
  })

  it('elo with multiple matches', () => {
    const playerRepo = new InMemoryPlayerRepository()
    playerRepo.save(player('p1', 'Alice'))
    playerRepo.save(player('p2', 'Bob'))
    playerRepo.save(player('p3', 'Charlie'))
    const gameTypeRepo = new InMemoryGameTypeRepository()
    gameTypeRepo.save(gameType('gt1', 'Test', 'HIGHEST_SCORE'))
    const matchRepo = new InMemoryMatchRepository()
    matchRepo.save(match('m1', 1000, 'gt1', [{ playerId: 'p1', score: 10 }, { playerId: 'p2', score: 5 }]))
    matchRepo.save(match('m2', 2000, 'gt1', [{ playerId: 'p1', score: 8 }, { playerId: 'p2', score: 12 }]))
    matchRepo.save(match('m3', 3000, 'gt1', [{ playerId: 'p3', score: 10 }, { playerId: 'p1', score: 3 }]))

    const result = buildUseCase(matchRepo, gameTypeRepo, playerRepo).invoke()
    const alice = result.find((r) => r.playerId === 'p1')!
    const bob = result.find((r) => r.playerId === 'p2')!
    const charlie = result.find((r) => r.playerId === 'p3')!

    expect(alice.elo).toBeLessThan(bob.elo)
    expect(bob.elo).toBeLessThan(charlie.elo)
    expect(result.map((r) => r.playerId)).toEqual(['p3', 'p2', 'p1'])
  })
})
